Pandas is a Python library or package. Which means that it is a coherent body of Python code that you can use in your Python code for a particular purpose.

The purpose of Pandas is to make working with "relational" or "labelled" data both easy and intuitive.

More precisely, and the following is taken from the Pandas documentation pages at GitHub, Pandas is well suited for many different kinds of data:

  - SQL tables or Excel spreadsheets (tabular data with heterogeneously-typed columns)
  - Time series data (ordered and unordered, not necessarily fixed-frequency)
  - Arbitrary matrix data (homo- or heterogeneously typed with row and column labels)
  - Other observational or statistical data sets (labelled or unlabelled)
  
Pandas is a good example of the "batteries included" principle of Python: Python comes with libraries, especially in the field of scientific computing, that make life a lot easier for it's users:

  - High-level code, often optimized for speed, to tackle common problems
  - Integration with other libraries
  
Pandas is built on top of another Python library, called Numpy. So some knowledge of that library will help us, users of Pandas, and in order to get a grasp of Numpy, some general knowledge of Python is necessary. The Russian doll principle, in short.

![Russian dolls](images/russian_dolls.jpeg)

Pandas is designed to do many things well, but here we will concentrate on some practical aspects of working with the library.

In order to be able to use a library one has to "load" that library, that is make the functions defined in the library  available to our programs.

There are two ways of enabling the use of the code contained in the libraries.

In [1]:
import pandas

But more often we see that import statement in another form:

In [2]:
import pandas as pd

The reason for this is that each time we want to use functions from the Pandas library in our code, we have to use the following syntax: LibraryName.FunctionName and pd.read_csv() is shorter (less typing) than it's equivalent pandas.read_csv.

To get a grip on Pandas, we start with a small ecological dataset from the US with survey data about animals caught on certain plots.

We already imported the Pandas library in one of the notebook cells above with the statement "import pandas as pd".

Now we have to get to the file. For this we need another library called os, that allows us to use commands that work on our operating system (os).

In [3]:
import os
os.getcwd()

'/Users/peter/Documents/bootcamps'

So, now we know where we are, the file we are going to work with is one directory below our "current working directory" or "cwd".

In [4]:
pd.read_csv("data/surveys.csv")

record_id  month  day  year  plot_id species_id  sex  hindfoot_length  \
0              1      7   16  1977        2         NL    M             32.0   
1              2      7   16  1977        3         NL    M             33.0   
2              3      7   16  1977        2         DM    F             37.0   
3              4      7   16  1977        7         DM    M             36.0   
4              5      7   16  1977        3         DM    M             35.0   
5              6      7   16  1977        1         PF    M             14.0   
6              7      7   16  1977        2         PE    F              NaN   
7              8      7   16  1977        1         DM    M             37.0   
8              9      7   16  1977        1         DM    F             34.0   
9             10      7   16  1977        6         PF    F             20.0   
10            11      7   16  1977        5         DS    F             53.0   
11            12      7   16  1977        7         DM    M             38.0   
12            13      7   16  1977        3         DM    M             35.0   
13            14      7   16  1977        8         DM  NaN              NaN   
14            15      7   16  1977        6         DM    F             36.0   
15            16      7   16  1977        4         DM    F             36.0   
16            17      7   16  1977        3         DS    F             48.0   
17            18      7   16  1977        2         PP    M             22.0   
18            19      7   16  1977        4         PF  NaN              NaN   
19            20      7   17  1977       11         DS    F             48.0   
20            21      7   17  1977       14         DM    F             34.0   
21            22      7   17  1977       15         NL    F             31.0   
22            23      7   17  1977       13         DM    M             36.0   
23            24      7   17  1977       13         SH    M             21.0   
24            25      7   17  1977        9         DM    M             35.0   
25            26      7   17  1977       15         DM    M             31.0   
26            27      7   17  1977       15         DM    M             36.0   
27            28      7   17  1977       11         DM    M             38.0   
28            29      7   17  1977       11         PP    M              NaN   
29            30      7   17  1977       10         DS    F             52.0   
...          ...    ...  ...   ...      ...        ...  ...              ...   
35519      35520     12   31  2002        9         SF  NaN             24.0   
35520      35521     12   31  2002        9         DM    M             37.0   
35521      35522     12   31  2002        9         DM    F             35.0   
35522      35523     12   31  2002        9         DM    F             36.0   
35523      35524     12   31  2002        9         PB    F             25.0   
35524      35525     12   31  2002        9         OL    M             21.0   
35525      35526     12   31  2002        8         OT    F             20.0   
35526      35527     12   31  2002       13         DO    F             33.0   
35527      35528     12   31  2002       13         US  NaN              NaN   
35528      35529     12   31  2002       13         PB    F             25.0   
35529      35530     12   31  2002       13         OT    F             20.0   
35530      35531     12   31  2002       13         PB    F             27.0   
35531      35532     12   31  2002       14         DM    F             34.0   
35532      35533     12   31  2002       14         DM    F             36.0   
35533      35534     12   31  2002       14         DM    M             37.0   
35534      35535     12   31  2002       14         DM    M             37.0   
35535      35536     12   31  2002       14         DM    F             35.0   
35536      35537     12   31  2002       14         DM    F             36.0   
35537  

Looks ok. Notice the missing values in some of the columns (NaN). Also notice that some columns, like plot and species reference ID's which makes this matrix a bit terse, although it is a sound way of doing things: Referencing entities stored in another table or matrix by way of ann ID.

In order to be able to work with the dataset surveys.csv, we need to keep the contents in memory. The way to do this, is to assign it to a new object by assigning the result to a variable. For this we use the syntactical element "=".

In [5]:
surveys_df = pd.read_csv("data/surveys.csv")

What precisely is this surveys_df thing? Let's ask Python:

In [6]:
type(surveys_df)

pandas.core.frame.DataFrame

Cool, let's see what is contained in the dataframe:

In [7]:
surveys_df

record_id  month  day  year  plot_id species_id  sex  hindfoot_length  \
0              1      7   16  1977        2         NL    M             32.0   
1              2      7   16  1977        3         NL    M             33.0   
2              3      7   16  1977        2         DM    F             37.0   
3              4      7   16  1977        7         DM    M             36.0   
4              5      7   16  1977        3         DM    M             35.0   
5              6      7   16  1977        1         PF    M             14.0   
6              7      7   16  1977        2         PE    F              NaN   
7              8      7   16  1977        1         DM    M             37.0   
8              9      7   16  1977        1         DM    F             34.0   
9             10      7   16  1977        6         PF    F             20.0   
10            11      7   16  1977        5         DS    F             53.0   
11            12      7   16  1977        7         DM    M             38.0   
12            13      7   16  1977        3         DM    M             35.0   
13            14      7   16  1977        8         DM  NaN              NaN   
14            15      7   16  1977        6         DM    F             36.0   
15            16      7   16  1977        4         DM    F             36.0   
16            17      7   16  1977        3         DS    F             48.0   
17            18      7   16  1977        2         PP    M             22.0   
18            19      7   16  1977        4         PF  NaN              NaN   
19            20      7   17  1977       11         DS    F             48.0   
20            21      7   17  1977       14         DM    F             34.0   
21            22      7   17  1977       15         NL    F             31.0   
22            23      7   17  1977       13         DM    M             36.0   
23            24      7   17  1977       13         SH    M             21.0   
24            25      7   17  1977        9         DM    M             35.0   
25            26      7   17  1977       15         DM    M             31.0   
26            27      7   17  1977       15         DM    M             36.0   
27            28      7   17  1977       11         DM    M             38.0   
28            29      7   17  1977       11         PP    M              NaN   
29            30      7   17  1977       10         DS    F             52.0   
...          ...    ...  ...   ...      ...        ...  ...              ...   
35519      35520     12   31  2002        9         SF  NaN             24.0   
35520      35521     12   31  2002        9         DM    M             37.0   
35521      35522     12   31  2002        9         DM    F             35.0   
35522      35523     12   31  2002        9         DM    F             36.0   
35523      35524     12   31  2002        9         PB    F             25.0   
35524      35525     12   31  2002        9         OL    M             21.0   
35525      35526     12   31  2002        8         OT    F             20.0   
35526      35527     12   31  2002       13         DO    F             33.0   
35527      35528     12   31  2002       13         US  NaN              NaN   
35528      35529     12   31  2002       13         PB    F             25.0   
35529      35530     12   31  2002       13         OT    F             20.0   
35530      35531     12   31  2002       13         PB    F             27.0   
35531      35532     12   31  2002       14         DM    F             34.0   
35532      35533     12   31  2002       14         DM    F             36.0   
35533      35534     12   31  2002       14         DM    M             37.0   
35534      35535     12   31  2002       14         DM    M             37.0   
35535      35536     12   31  2002       14         DM    F             35.0   
35536      35537     12   31  2002       14         DM    F             36.0   
35537  

To generate a smaller preview use the head() and tail() methods:

In [8]:
surveys_df.head()

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2          3      7   16  1977        2         DM   F             37.0   
3          4      7   16  1977        7         DM   M             36.0   
4          5      7   16  1977        3         DM   M             35.0   

   weight  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

The Python slicing syntax also works on dataframes:

In [9]:
surveys_df[2:4]

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
2          3      7   16  1977        2         DM   F             37.0   
3          4      7   16  1977        7         DM   M             36.0   

   weight  
2     NaN  
3     NaN

There are various ways to peek into the object to see precisely what we are dealing with.

We can use the predefined and general method "info" on our dataframe object:

In [10]:
surveys_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35549 entries, 0 to 35548
Data columns (total 9 columns):
record_id          35549 non-null int64
month              35549 non-null int64
day                35549 non-null int64
year               35549 non-null int64
plot_id            35549 non-null int64
species_id         34786 non-null object
sex                33038 non-null object
hindfoot_length    31438 non-null float64
weight             32283 non-null float64
dtypes: float64(2), int64(5), object(2)
memory usage: 2.4+ MB


This gives us a lot of information:

- the type of the object: <class 'pandas.core.frame.DataFrame'>
- The number of entries (rows) in the dataframe
- The number of data columns: 9
- The number of missing values per column, if any: sex = 35549 - 33038 => 2511 (7%)
- The types (dtypes) of data contained in the matrix: int64, object, float64

Info() method gives an overview of the dataframe, but one can also directly ask for the types of the data:

In [11]:
surveys_df.dtypes

record_id            int64
month                int64
day                  int64
year                 int64
plot_id              int64
species_id          object
sex                 object
hindfoot_length    float64
weight             float64
dtype: object

Dataframes also have a describe method, that provides us with some basic statistics about the datasets numeric columns:

In [12]:
surveys_df.describe()

/Users/peter/anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


record_id         month           day          year       plot_id  \
count  35549.000000  35549.000000  35549.000000  35549.000000  35549.000000   
mean   17775.000000      6.474022     16.105966   1990.475231     11.397001   
std    10262.256696      3.396583      8.256691      7.493355      6.799406   
min        1.000000      1.000000      1.000000   1977.000000      1.000000   
25%     8888.000000      4.000000      9.000000   1984.000000      5.000000   
50%    17775.000000      6.000000     16.000000   1990.000000     11.000000   
75%    26662.000000      9.000000     23.000000   1997.000000     17.000000   
max    35549.000000     12.000000     31.000000   2002.000000     24.000000   

       hindfoot_length        weight  
count     31438.000000  32283.000000  
mean         29.287932     42.672428  
std           9.564759     36.631259  
min           2.000000      4.000000  
25%                NaN           NaN  
50%                NaN           NaN  
75%                NaN           NaN  
max          70.000000    280.000000

Another way of thinking of dataframes is as a group of series that share an index, the column headers. We can easily select a column, using the following syntax:

In [13]:
surveys_df['weight'].head()

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
Name: weight, dtype: float64

Using column selection is an easy way to select some columns and place them in a new dataframe:

In [14]:
smaller_surveys_df = surveys_df[['record_id', 'plot_id', 'weight']]
smaller_surveys_df.head()

record_id  plot_id  weight
0          1        2     NaN
1          2        3     NaN
2          3        2     NaN
3          4        7     NaN
4          5        3     NaN

Of course, we can select rows from our dataframe too:

In [15]:
print(surveys_df[surveys_df.weight > 100].head(5))

     record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
356        357     11   12  1977        9         DS   F             50.0   
361        362     11   12  1977        1         DS   F             51.0   
366        367     11   12  1977       20         DS   M             51.0   
376        377     11   12  1977        9         DS   F             48.0   
380        381     11   13  1977       17         DS   F             48.0   

     weight  
356   117.0  
361   121.0  
366   115.0  
376   120.0  
380   118.0  


The data we have been working with sofar, surveys.csv, contains two columns that reference data available in other tables (csv files). The columns "plot_id" and "species_id" are made up of id's (identifiers) that identify rows in two additional files: plots.csv and species.csv.

Let's load the two files into Pandas as dataframes and have a look at them:

In [20]:
species_df = pd.read_csv("data/species.csv")
species_df.head()

species_id             genus          species    taxa
0         AB        Amphispiza        bilineata    Bird
1         AH  Ammospermophilus          harrisi  Rodent
2         AS        Ammodramus       savannarum    Bird
3         BA           Baiomys          taylori  Rodent
4         CB   Campylorhynchus  brunneicapillus    Bird

Here we see, just as in our initial dataframe, a column with the species_id and some other columns that contain data about the species.

Combining two datasets on the basis of a shared ID is often referred to as a join. In Pandas, however, the method is called "merge":

In [22]:
pd.merge(surveys_df, species_df, on="species_id", how="inner").head(5)

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2         22      7   17  1977       15         NL   F             31.0   
3         38      7   17  1977       17         NL   M             33.0   
4         72      8   19  1977        2         NL   M             31.0   

   weight    genus   species    taxa  
0     NaN  Neotoma  albigula  Rodent  
1     NaN  Neotoma  albigula  Rodent  
2     NaN  Neotoma  albigula  Rodent  
3     NaN  Neotoma  albigula  Rodent  
4     NaN  Neotoma  albigula  Rodent